### DISCUSSION TITLE
Verification codes other than the first one are not working

### DISCUSSION THREAD

**Discussion by MixaKonan** [aws/aws-sdk-net-extensions-cognito#138] [Q&A] [answered]
### Describe the bug

I'm not actually sure whether this is SDK or Cognito problem or maybe I just missed something, but any help would be much appreciated.

I have a user pool set up in AWS Cognito. I've created a user and made it confirmed be changing the password via `RespondToNewPasswordRequiredAsync`.

I'm trying to implement 'Forgot Password' feature in my system. After making a 'ForgotPasswordAsync' I successfully receive a verification code.

However, I also try to implement a 'Resend verification code' feature. And whilst I do receive a new code, I get an 'Invalid verification code' exception. And this happens for every code other than the first one.

![image](https://github.com/aws/aws-sdk-net-extensions-cognito/assets/37301415/2b6d4a90-4045-48f9-828f-3b7e2ac93618)

I thought that maybe there is some time that enables new codes, but I've tried to wait 1, 5 and 10 minutes and no success was accomplished.

As I said, the first one received is working with no issues whatsoever.

![image](https://github.com/aws/aws-sdk-net-extensions-cognito/assets/37301415/53027f70-91ca-4cca-8338-b65e4feab9f2)

### Expected Behavior

Any received code should work or at least the latest one.

### Current Behavior

Only the first code received works

### Reproduction Steps

1. Create a user pool in AWS Cognito.
2. Create a new user with verified email.
3. Change this user's password so it becomes 'Confirmed'.
4. Make a 'Forgot Password' request with SDK.
5. Receive a code but ignore it
6. Make a 'Resend confirmation code' request with SDK.
7. Make a 'Confirm Forgot Password' request with the latest code received

### Possible Solution

No solutions were found

### AWS .NET SDK and/or Package version used

Amazon.Extensions.CognitoAuthentication: 2.5.2
AWSSDK.CognitoIdentity: 3.7.300.10
AWSSDK.CognitoIdentityProvider: 3.7.300.10
AWSSDK.Core: 3.7.300.10

### Targeted .NET Platform

.NET 6

### Operating System and version

Windows 10, AmazonLinux (docker container)

**Comment by ashishdhingra** [comment]
Needs investigation.

**Comment by ashishdhingra** [comment, marked as answer]
@MixaKonan Good afternoon. I'm not sure if you are using the correct logic. The [ConfirmForgotPassword](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ConfirmForgotPassword.html) is the service API operation invoked by [CognitoUser.ConfirmForgotPasswordAsync()](https://github.com/aws/aws-sdk-net-extensions-cognito/blob/4d6c4af4b751c5f67698790199c18aa80e45272a/src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs#L244C29-L244C55).

How are you sending the verification code?
- Are you using [CognitoUser.ForgotPasswordAsync()](https://github.com/aws/aws-sdk-net-extensions-cognito/blob/4d6c4af4b751c5f67698790199c18aa80e45272a/src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs#L220C29-L220C48), which invokes [ForgotPassword](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ForgotPassword.html) service API call, OR
- Are you using [CognitoUser.ResendConfirmationCodeAsync()](https://github.com/aws/aws-sdk-net-extensions-cognito/blob/4d6c4af4b751c5f67698790199c18aa80e45272a/src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs#L199), which invokes service API operation [ResendConfirmationCode](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ResendConfirmationCode.html).

The [ResendConfirmationCode](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ResendConfirmationCode.html) API call `Resends the confirmation (for confirmation of registration) to a specific user in the user pool.`. In your use case, user is **already registered**. So `ResendConfirmationCode` is not valid for forgot password scenario. You should try re-invoking the [CognitoUser.ForgotPasswordAsync()](https://github.com/aws/aws-sdk-net-extensions-cognito/blob/4d6c4af4b751c5f67698790199c18aa80e45272a/src/Amazon.Extensions.CognitoAuthentication/CognitoUser.cs#L220C29-L220C48) to resend **forgot password** verification code and it should resend the code (if Cognito supports it). Please let me know if it works.

Thanks,
Ashish

**Comment by MixaKonan** [comment]
@ashishdhingra Good afternoon to you too. Indeed you were correct in your assumptions and I was actually using incorrect flow. I missed the difference between 'confirmation code' and 'password reset code' and it's actually stated in the documentation that 'ResendConfirmationCode' is used for registration only.

>Resends the confirmation (for confirmation of registration) to a specific user in the user pool.

The issue is resolved and I will close it. Thank you very much for your support and the time wasted.

[a comment by an automated account is not shown]
